torch/fx/experimental/proxy_tensor.py

```diff
@@ -51,10 +51,18 @@ def decompose(decomposition_table):
 proxy_slot = object()
 no_default = object()
 
-def set_proxy_slot(obj, tracer, proxy):
-    d = obj.__dict__.setdefault(proxy_slot, weakref.WeakKeyDictionary())
-    assert isinstance(d, weakref.WeakKeyDictionary)
-    d[tracer] = proxy
+# We cannot use __dict__ on tensor for nested tensors
```

Comment:
Why not. This seems like a nested tensor problem.

Reply:
Yes - the nested_tensor impl has issues. 

I was hoping to work around it, but its too deep rooted, I am looking at maybe just fixing the underlying problems in nested_tensor.

2 examples:

**1. ID instability:**

```
a, b = torch.tensor([0.5, 0.5]), torch.tensor([0.33, 0.33])
nn = torch.nested_tensor([a, b])
for x in nn:
    print(id(x))

print("---")
for x in nn:
    print(id(x))
```
Output
```
139763614730016
139763573954000
---
139763614730016
139763573954960
```

**2. Dict instability:**
```
print(nn[0].__dict__)
    nn[0].__dict__['voz2'] = 'Test2'
    print(nn[0].__dict__)
```
Output
```
{}
{}
```

I was hoping nested_tensor was more like a list, but it's got a bit of nuance around how it makes a NestedNode<at::Tensor>.

I have not read it in depth yet, but the first place that looks potentially suspicious is:

`flat_tensors.push_back(tensor_node.children(i).reshape(-1).contiguous());` 

Where tensor_node.children() is the aten::tensor object originally passed to the nested_tensor() ctor.

The underlying cause seems to be that we are not correctly associating the PyObject of the py-side of the tensor with these tensors - we are probably copying somewhere on iteration...

Reply:
Oh, this is probably the thing where nested tensor did its Python bindings from scratch, rather than directly using Tensor. The correct fix is to directly use Tensor.

@drisspg @cpuhrsch IIRC, you guys introduced your own Python bindings so you could override the behavior of a few methods like size(). If we let you define a tensor subclass for NestedTensor that subclasses Tensor, and gets automatically used when a NestedTensor goes to Python, would that suffice?

Reply:
I don't think that this is the case where we implemented python bindings from scratch, and more related to the nestedtensorimpl in c++. I was speaking with Michael Voznesensky about this yesterday and he seemed to have many opinions on the matter.

Reply:
@ezyang - we did not land custom python bindings. `nested_tensor` is a native function and uses the usual Python binding code gen.

It's a regular torch.Tensor backed by a NestedTensorImpl (similar to our sparse Tensors).

Reply:
@voznesenskym - On your concerns around dict and id instability, here are the results for a regular Tensor.

```
>>> nn = torch.tensor([[0.5, 0.5], [0.33, 0.33]])
>>> for x in nn:
...     print(id(x))
...
140380897135728
140380897135888
>>> for x in nn:
...     print(id(x))
...
140380897135728
140380897135968
>>> nn[0].__dict__
{}
>>> nn[0].__dict__['voz2'] = 'Test2'
>>> nn[0].__dict__
{}
```

It's the same behavior. NestedTensor does not aim to be a Python list, it's a generalization of Tensors in shape.

Reply:
oh, I read the original example too fast

@voznesenskym whenever an unbind (aka iteration over tensor) happens, you need to trace it separately; i.e. fresh proxies must be created each time.